src/core/experimental/array.d

```diff
@@ -0,0 +1,2559 @@
+        else                                    alias Unqual =        T;
+    }
+    else // workaround
+    {
+             static if (is(T U ==          immutable U)) alias Unqual = U;
+        else static if (is(T U == shared inout const U)) alias Unqual = U;
+        else static if (is(T U == shared inout       U)) alias Unqual = U;
+        else static if (is(T U == shared       const U)) alias Unqual = U;
+        else static if (is(T U == shared             U)) alias Unqual = U;
+        else static if (is(T U ==        inout const U)) alias Unqual = U;
+        else static if (is(T U ==        inout       U)) alias Unqual = U;
+        else static if (is(T U ==              const U)) alias Unqual = U;
+        else                                             alias Unqual = T;
+    }
+}
+
+/**
+Is `From` implicitly convertible to `To`?
+ */
+enum bool isImplicitlyConvertible(From, To) = is(From : To);
+
+/**
+The element type of `R`. `R` does not have to be a range. The element type is
+determined as the type yielded by `r.front` for an object `r` of type `R`. For
+example, `ElementType!(T[])` is `T` if `T[]` isn't a narrow string; if it is, the
+element type is `dchar`. If `R` doesn't have `front`, `ElementType!R` is `void`.
+ */
+template ElementType(R)
+{
+    static if (is(typeof(R.init.front.init) T))
+        alias ElementType = T;
+    else
+        alias ElementType = void;
+}
+
+/**
+Yields `true` if `R` has a `length` member that returns a value of `size_t`
+type. `R` does not have to be a range. If `R` is a range, algorithms in the
+standard library are only guaranteed to support `length` with type `size_t`.
+*/
+template hasLength(R)
+{
+    static if (is(typeof(((R* r) => r.length)(null)) Length))
+        enum bool hasLength = is(Length == size_t);
+        //TODO: enum bool hasLength = is(Length == size_t) && !isNarrowString!R;
+    else
+        enum bool hasLength = false;
+}
+
+// { Allocators
+
+/**
+Returns the size in bytes of the state that needs to be allocated to hold an
+object of type `T`. `stateSize!T` is zero for `struct`s that are not
+nested and have no nonstatic member variables.
+ */
+template stateSize(T)
+{
+    static if (is(T == class) || is(T == interface))
+        enum stateSize = __traits(classInstanceSize, T);
+    else static if (is(T == void))
+        enum size_t stateSize = 0;
+    else
+        enum stateSize = T.sizeof;
+}
+
+template isAbstractClass(T...)
+if (T.length == 1)
+{
+    enum bool isAbstractClass = __traits(isAbstractClass, T[0]);
+}
+
+template isInnerClass(T)
+if (is(T == class))
+{
+    static if (is(typeof(T.outer)))
+        enum isInnerClass = __traits(isSame, typeof(T.outer), __traits(parent, T));
+    else
+        enum isInnerClass = false;
+}
+
+enum classInstanceAlignment(T) = size_t.alignof >= T.alignof ? size_t.alignof : T.alignof;
+
+T emplace(T, Args...)(T chunk, auto ref Args args)
+if (is(T == class))
```

Comment:
Needs to handle non D classes.

Reply:
Got it. What non D classes are supported? C++ ones?

Reply:
@jacob-carlborg that was copied from the phobos implementation (later to be moved), can you explain what needs to be done here? Thanks! Also: would rather add that as an additional PR after this.

Reply:
TypeInfo is somewhat broken for non D classes. But `emplace` might not be affected.

Reply:
Should be private then.

Reply:
What do you guys think about moving `emplace` to `core.internal.traits` ?

Reply:
It's not really a trait is it?

Reply:
You are right. My bad, sorry :).
The question remains: should we move it into a `core.internal.conv` ?

I think `emplace` is especially usefull when you want to construct an immutable object into a predefined memory area.

Reply:
Let's make it private for now and postpone that decision. Fight one fight at a time.